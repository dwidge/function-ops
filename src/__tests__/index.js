require("jest-specific-snapshot");
const instrument = require("../index");

const snapshotSrc = "__snapshots__/index.js.src.txt";

const makeLogger = (f) =>
  jest.fn((type, pos, ...args) => {
    //expect(snippet(f, pos)).toEqual(args[0][0]);
  });

const snippet = (f, pos) => f.toString().slice(...pos);

test("binary", async () => {
  const f = function (a, b, c) {
    a = b + c * 2;
    return a;
  };

  const logger = makeLogger(f);
  const logging = instrument(f, logger);
  expect(await logging(3, 4, 5)).toBe(14);
  expect(logger.mock.calls).toMatchSnapshot();
  expect(logging.toString()).toMatchSpecificSnapshot(snapshotSrc);
});

test("escape string", async () => {
  const f = function (a, b, c) {
    return "x'x" + a * b + c + 'z"z';
  };

  const logger = makeLogger(f);
  const logging = instrument(f, logger);
  expect(await logging(3, 4, "y")).toBe("x'x12yz\"z");
  expect(logger.mock.calls).toMatchSnapshot();
  expect(logging.toString()).toMatchSpecificSnapshot(snapshotSrc);
});

test("call", async () => {
  const f = function (a, b, c) {
    const d = (x, y) => x ** y;
    return Math.pow(4 * a, 2) + d(a, b) + c;
  };

  const logger = makeLogger(f);
  const logging = instrument(f, logger);
  expect(await logging(3, 4, 5)).toBe(230);
  expect(logger.mock.calls).toMatchSnapshot();
  expect(logging.toString()).toMatchSpecificSnapshot(snapshotSrc);
});

test("async", async () => {
  const f = async function (a, b, c) {
    const d = async (x, y) => x ** y;
    return (await d(a, b)) + c;
  };

  const logger = makeLogger(f);
  const logging = instrument(f, logger);
  expect(await logging(3, 4, 5)).toBe(86);
  expect(logger.mock.calls).toMatchSnapshot();
  expect(logging.toString()).toMatchSpecificSnapshot(snapshotSrc);
});

test("for loop", async () => {
  const f = async function (a, b, c) {
    for (var i = 0; i < a + b; i++) c = c + i;
    return c;
  };

  const logger = makeLogger(f);
  const logging = instrument(f, logger);
  expect(await logging(3, 4, 5)).toBe(26);
  expect(logger.mock.calls).toMatchSnapshot();
  expect(logging.toString()).toMatchSpecificSnapshot(snapshotSrc);
});

test("assignment", async () => {
  const f = function (a, b, c) {
    c += a - b;
    return c;
  };

  const logger = makeLogger(f);
  const logging = instrument(f, logger);
  expect(await logging(3, 4, 5)).toBe(4);
  expect(logger.mock.calls).toMatchSnapshot();
  expect(logging.toString()).toMatchSpecificSnapshot(snapshotSrc);
});
