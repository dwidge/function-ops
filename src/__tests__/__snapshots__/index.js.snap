// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`assignment 1`] = `
[
  [
    "BinaryExpression",
    [
      30,
      35,
    ],
    [
      "a - b",
      -1,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
]
`;

exports[`async 1`] = `
[
  [
    "BinaryExpression",
    [
      57,
      63,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      57,
      63,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      76,
      95,
    ],
    [
      "(await d(a, b)) + c",
      86,
    ],
    [
      "await d(a, b)",
      81,
    ],
    [
      "c",
      5,
    ],
  ],
  [
    "BinaryExpression",
    [
      57,
      63,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "CallExpression",
    [
      83,
      90,
    ],
    [
      "d(a, b)",
      81,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      57,
      63,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
]
`;

exports[`binary 1`] = `
[
  [
    "BinaryExpression",
    [
      29,
      38,
    ],
    [
      "b + c * 2",
      14,
    ],
    [
      "b",
      4,
    ],
    [
      "c * 2",
      10,
    ],
  ],
  [
    "BinaryExpression",
    [
      33,
      38,
    ],
    [
      "c * 2",
      10,
    ],
    [
      "c",
      5,
    ],
    [
      "2",
      2,
    ],
  ],
]
`;

exports[`call 1`] = `
[
  [
    "BinaryExpression",
    [
      45,
      51,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      45,
      51,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      64,
      96,
    ],
    [
      "Math.pow(4 * a, 2) + d(a, b) + c",
      230,
    ],
    [
      "Math.pow(4 * a, 2) + d(a, b)",
      225,
    ],
    [
      "c",
      5,
    ],
  ],
  [
    "BinaryExpression",
    [
      45,
      51,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      45,
      51,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      64,
      92,
    ],
    [
      "Math.pow(4 * a, 2) + d(a, b)",
      225,
    ],
    [
      "Math.pow(4 * a, 2)",
      144,
    ],
    [
      "d(a, b)",
      81,
    ],
  ],
  [
    "CallExpression",
    [
      64,
      82,
    ],
    [
      "Math.pow(4 * a, 2)",
      144,
    ],
    [
      "4 * a",
      12,
    ],
    [
      "2",
      2,
    ],
  ],
  [
    "BinaryExpression",
    [
      73,
      78,
    ],
    [
      "4 * a",
      12,
    ],
    [
      "4",
      4,
    ],
    [
      "a",
      3,
    ],
  ],
  [
    "BinaryExpression",
    [
      45,
      51,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
  [
    "CallExpression",
    [
      85,
      92,
    ],
    [
      "d(a, b)",
      81,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      45,
      51,
    ],
    [
      "x ** y",
      81,
    ],
    [
      "x",
      3,
    ],
    [
      "y",
      4,
    ],
  ],
]
`;

exports[`escape string 1`] = `
[
  [
    "BinaryExpression",
    [
      32,
      57,
    ],
    [
      ""x'x" + a * b + c + 'z"z'",
      "x'x12yz"z",
    ],
    [
      ""x'x" + a * b + c",
      "x'x12y",
    ],
    [
      "'z"z'",
      "z"z",
    ],
  ],
  [
    "BinaryExpression",
    [
      32,
      49,
    ],
    [
      ""x'x" + a * b + c",
      "x'x12y",
    ],
    [
      ""x'x" + a * b",
      "x'x12",
    ],
    [
      "c",
      "y",
    ],
  ],
  [
    "BinaryExpression",
    [
      32,
      45,
    ],
    [
      ""x'x" + a * b",
      "x'x12",
    ],
    [
      ""x'x"",
      "x'x",
    ],
    [
      "a * b",
      12,
    ],
  ],
  [
    "BinaryExpression",
    [
      40,
      45,
    ],
    [
      "a * b",
      12,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
]
`;

exports[`for loop 1`] = `
[
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      0,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      5,
    ],
    [
      "c",
      5,
    ],
    [
      "i",
      0,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      1,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      6,
    ],
    [
      "c",
      5,
    ],
    [
      "i",
      1,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      2,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      8,
    ],
    [
      "c",
      6,
    ],
    [
      "i",
      2,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      3,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      11,
    ],
    [
      "c",
      8,
    ],
    [
      "i",
      3,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      4,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      15,
    ],
    [
      "c",
      11,
    ],
    [
      "i",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      5,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      20,
    ],
    [
      "c",
      15,
    ],
    [
      "i",
      5,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      true,
    ],
    [
      "i",
      6,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
  [
    "BinaryExpression",
    [
      67,
      72,
    ],
    [
      "c + i",
      26,
    ],
    [
      "c",
      20,
    ],
    [
      "i",
      6,
    ],
  ],
  [
    "BinaryExpression",
    [
      47,
      56,
    ],
    [
      "i < a + b",
      false,
    ],
    [
      "i",
      7,
    ],
    [
      "a + b",
      7,
    ],
  ],
  [
    "BinaryExpression",
    [
      51,
      56,
    ],
    [
      "a + b",
      7,
    ],
    [
      "a",
      3,
    ],
    [
      "b",
      4,
    ],
  ],
]
`;
